diff --git a/package/network/services/hostapd/files/hostapd.sh b/package/network/services/hostapd/files/hostapd.sh
index 34e4bdb..d85059b 100644
--- a/package/network/services/hostapd/files/hostapd.sh
+++ b/package/network/services/hostapd/files/hostapd.sh
@@ -226,7 +226,10 @@ hostapd_common_add_bss_config() {
 	config_add_boolean wds wmm uapsd hidden utf8_ssid
 
 	config_add_int maxassoc max_inactivity
-	config_add_boolean disassoc_low_ack isolate short_preamble
+	config_add_boolean disassoc_low_ack isolate short_preamble low_rssi_disconnect
+
+	config_add_int signal_connect signal_stay signal_poll_time \
+	signal_drop_reason signal_strikes
 
 	config_add_int \
 		wep_rekey eap_reauth_period \
@@ -322,7 +325,8 @@ hostapd_set_bss_options() {
 		iapp_interface eapol_version dynamic_vlan ieee80211w nasid \
 		acct_server acct_secret acct_port acct_interval \
 		bss_load_update_period chan_util_avg_period sae_require_mfp \
-		multi_ap multi_ap_backhaul_ssid multi_ap_backhaul_key
+		multi_ap multi_ap_backhaul_ssid multi_ap_backhaul_key \
+		signal_connect signal_stay signal_poll_time signal_drop_reason signal_strikes low_rssi_disconnect
 
 	set_default isolate 0
 	set_default maxassoc 0
@@ -340,6 +344,12 @@ hostapd_set_bss_options() {
 	set_default chan_util_avg_period 600
 	set_default utf8_ssid 1
 	set_default multi_ap 0
+	set_default low_rssi_disconnect 0
+	set_default signal_connect -128
+	set_default signal_stay -128
+	set_default signal_poll_time 10
+	set_default signal_drop_reason 3
+	set_default signal_strikes 3
 
 	append bss_conf "ctrl_interface=/var/run/hostapd"
 	if [ "$isolate" -gt 0 ]; then
@@ -361,6 +371,15 @@ hostapd_set_bss_options() {
 	append bss_conf "uapsd_advertisement_enabled=$uapsd" "$N"
 	append bss_conf "utf8_ssid=$utf8_ssid" "$N"
 	append bss_conf "multi_ap=$multi_ap" "$N"
+	append bss_conf "low_rssi_disconnect=$low_rssi_disconnect" "$N"
+
+	if [ "$low_rssi_disconnect" -gt 0 ]; then
+			append bss_conf "signal_connect=$signal_connect" "$N"
+			append bss_conf "signal_stay=$signal_stay" "$N"
+			append bss_conf "signal_poll_time=$signal_poll_time" "$N"
+			append bss_conf "signal_strikes=$signal_strikes" "$N"
+			append bss_conf "signal_drop_reason=$signal_drop_reason" "$N"
+	fi
 
 	[ "$tdls_prohibit" -gt 0 ] && append bss_conf "tdls_prohibit=$tdls_prohibit" "$N"
 
